### PR TITLE
Aa add mapping options

### PR DESCRIPTION
**Pull request by yodarjun**
This adds the ability to set `_parent` and `_routing` options on mappings.

### CODE REVIEW

File: lib/elastic_model.rb [hunk 1]
```diff
@@ -8,4 +8,4 @@
 
 # TODO: error out if host is not defined
 log = ENV['debug'] ? true : false
-$es ||= Elasticsearch::Client.new(:host => 'localhost:9200', :log => log)
+$es ||= Elasticsearch::Client.new(:host => 'localhost:19200', :log => log)
```
Comment (quentindecock) [inline lib/elastic_model.rb]: You'd rather use the Boxen env variable for this and || it with the default ES port

Comment (kylessnell) [inline lib/elastic_model.rb]: was going to say the same thing. if this is released as a gem, non boxen users will use 9200 by default.

File: spec/lib/elastic_model/integration/callbacks_spec.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper'
 
 describe ElasticModel::Callbacks do
-  context 'persistance' do
+  describe 'persistance' do
```
Comment (benseligman) [inline spec/lib/elastic_model/integration/callbacks_spec.rb]: persistence

File: lib/elastic_model/instrumentation.rb [hunk 3]
```diff
@@ -62,17 +82,38 @@ def self.mapping_for(field_name, options = {})
...
     def save_to_es!
-      $es.index :index => self.class.es_index_name, :type => self.class.es_type, :id => self.id, :body => self.as_json
+      body   = self.as_json
+      params = {
+        :index => self.class.es_index_name,
+        :type  => self.class.es_type,
+        :id    => self.id,
+        :body  => body
+      }
+      if self.class.es_parent_present?
```
Comment (quentindecock) [inline lib/elastic_model/instrumentation.rb]: This is going to be slow as fuck. As everytime you are going to write data to elastic search, you will do a get mapping. This is a no go...

File: spec/config/mongoid.yml [hunk 4]
```diff
@@ -3,4 +3,4 @@ development:
     default:
       database: elastic_model
       hosts:
-        - localhost:27017
+        - localhost:17017
```
Comment (quentindecock) [inline spec/config/mongoid.yml]: Same thing here. Use the Boxen environment variable.

File: lib/elastic_model/instrumentation.rb (hunk 3 above)
Comment (quentindecock) [inline lib/elastic_model/instrumentation.rb]: Also, how are we going to use bulk imports with this new scheme?

File: lib/elastic_model.rb (hunk 1 above)
Comment (michaeldauria) [inline lib/elastic_model.rb]:
```
$es ||= Elasticsearch::Client.new(:host => ENV['BOXEN_ELASTICSEARCH_URL'] || 'localhost:9200', :log => log)
```

**Comment by benseligman** [comment]
lgtm after quentin's comments

**Comment by michaeldauria** [comment]
I have to ask, why not just move to https://github.com/elasticsearch/elasticsearch-rails ?

Might as well rely on the library that is maintained by the ES folks themselves.

After the small changes, :+1:

File: lib/elastic_model/instrumentation.rb (hunk 3 above)
Comment (yodarjun) [inline lib/elastic_model/instrumentation.rb]: @quentindecock im not sure about bulk imports. we can discuss the implications when you have some time

**Comment by benseligman** [comment]
:+1:

**Comment by kylessnell** [comment]
:+1:

File: lib/elastic_model/instrumentation.rb [hunk 5]
```diff
@@ -62,17 +84,36 @@ def self.mapping_for(field_name, options = {})
...
+      params = {
+        :index => self.class.es_index_name,
+        :type  => self.class.es_type,
+        :id    => self.id,
+        :body  => body
+      }
+      if self.class.has_es_parent?
+        begin
+          params[:parent]  = self.es_parent_id
+          params[:routing] = self.es_parent_id
```
Comment (quentindecock) [inline lib/elastic_model/instrumentation.rb]: According to the article I sent to you, this is un-necessary. If set as a mapping, Elasticsearch should load the document, extract the routing and set it up, serialize the document and then post it an Elasticsearch index.

This is how you set it up:

```
$ curl -XPUT 'http://localhost:9200/store/order/_mapping' -d '
{
   "order":{
      "_routing":{
         "required":true,
         "path":"customerID"
      }
   }
}'
```

File: spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      end
+    end
+    let!(:parent_instance) { parent_class.create }
+
+    let!(:valid_associated_class) do
+      define_constant('valid_associated_class') do
+        include Mongoid::Document
+        include ElasticModel::Instrumentation
+        include ElasticModel::Callbacks
+        create_es_index
```
Comment (quentindecock) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:24]: remove `create_es_index` here you call it below.

File: spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+        field :count, :type => Integer
+        belongs_to :parent_association
+        es_index_name "test_classes"
```
Comment (quentindecock) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:28]: Do you intend on creating there in the same index? Why not leaving `es_index_name` and `es_type` blank?

File: spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
+require 'spec_helper'
+
+describe ElasticModel::Callbacks do
+  describe "saving indexes with associations" do
```
Comment (quentindecock) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:4]: I do not understand the title. This is unclear what the intent is.

**Pull request closed without merging**